src/textual/widgets/_data_table.py

```diff
@@ -26,40 +29,125 @@
 from ..scroll_view import ScrollView
 from ..strip import Strip
 
+CellCacheKey: TypeAlias = "tuple[RowKey, ColumnKey, Style, bool, bool, int]"
+LineCacheKey: TypeAlias = (
+    "tuple[int, int, int, int, Coordinate, Coordinate, Style, CursorType, bool, int]"
+)
+RowCacheKey: TypeAlias = (
+    "tuple[RowKey, int, Style, Coordinate, Coordinate, CursorType, bool, bool, int]"
+)
 CursorType = Literal["cell", "row", "column", "none"]
-CELL: CursorType = "cell"
 CellType = TypeVar("CellType")
 
 
 class CellDoesNotExist(Exception):
-    pass
+    """The cell key/index was invalid.
 
+    Raised when the user supplies coordinates or cell keys which
+    do not exist in the DataTable."""
 
-def default_cell_formatter(obj: object) -> RenderableType | None:
-    """Format a cell in to a renderable.
+
+class DuplicateKey(Exception):
+    """The key supplied already exists.
+
+    Raised when the RowKey or ColumnKey provided already refers to
+    an existing row or column in the DataTable. Keys must be unique."""
+
+
+@functools.total_ordering
+class StringKey:
+    """An object used as a key in a mapping.
+
+    It can optionally wrap a string,
+    and lookups into a map using the object behave the same as lookups using
+    the string itself."""
+
+    value: str | None
+
+    def __init__(self, value: str | None = None):
+        self.value = value
+
+    def __hash__(self):
+        # If a string is supplied, we use the hash of the string. If no string was
+        # supplied, we use the default hash to ensure uniqueness amongst instances.
+        return hash(self.value) if self.value is not None else id(self)
+
+    def __eq__(self, other: object) -> bool:
+        # Strings will match Keys containing the same string value.
+        # Otherwise, you'll need to supply the exact same key object.
+        if isinstance(other, (str, StringKey)):
+            return hash(self) == hash(other)
```

Comment:
There's a tiny, but non-zero chance that comparing hashes may return false positives.

I think we need to use string comparison if both sides have a str.

Something like this: 

```suggestion
        if isinstance(other, str):
            return self.value == other
        elif isinstance(other, StringKey):
            if self.value is not None and other.value is not None:
                return self.value == other.value
            else:
                return hash(self) == hash(other)               
```